File: Exercicios-LPA-main/Aula 21/ex.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <errno.h>

// Função para criar um diretório
static int criar_diretorio(const char *nome_dir) {
    // Tentativa de criar o diretório
    if (mkdir(nome_dir, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) == -1) {
        fprintf(stderr, "Erro ao criar diretório '%s': %s\n", 
                nome_dir, strerror(errno));
        return 0;
    }
    printf("Diretório '%s' criado com sucesso\n", nome_dir);
    return 1;
}

// Função para listar conteúdo do diretório atual
static void listar_diretorio(void) {
    DIR *dir;
    struct dirent *ent;

    // Tentativa de abrir o diretório atual
    dir = opendir(".");
    if (dir == NULL) {
        fprintf(stderr, "Erro ao abrir diretório atual: %s\n", 
                strerror(errno));
        return;
    }

    printf("Conteúdo do diretório atual:\n");
    
    // Iteração sobre os itens do diretório
    while ((ent = readdir(dir)) != NULL) {
        printf("%s\n", ent->d_name);
    }

    closedir(dir);
}

int main(int argc, char *argv[]) {
    int i;
    
    // Verificação se há argumentos suficientes
    if (argc < 2) {
        fprintf(stderr, "Uso: %s [-c nome_dir] [-l]\n", argv[0]);
        return EXIT_FAILURE;
    }

    // Processamento dos argumentos
    for (i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-c") == 0 && i + 1 < argc) {
            criar_diretorio(argv[++i]);
        } else if (strcmp(argv[i], "-l") == 0) {
            listar_diretorio();
        }
    }

    return EXIT_SUCCESS;
}

File: Exercicios-LPA-main/Aula 21/ex.py
import os
from pathlib import Path
import argparse
from typing import Optional

def criar_diretorio(nome_dir: str) -> None:
    """
    Cria um novo diretório com tratamento de erros.
    
    Args:
        nome_dir: Nome do diretório a ser criado
        
    Raises:
        OSError: Se ocorrer erro na criação do diretório
    """
    try:
        os.mkdir(nome_dir)
        print(f"Diretório '{nome_dir}' criado com sucesso")
    except FileExistsError:
        print(f"Erro: O diretório '{nome_dir}' já existe")
    except PermissionError:
        print(f"Erro: Permissão negada para criar o diretório '{nome_dir}'")
    except OSError as e:
        print(f"Erro ao criar diretório '{nome_dir}': {str(e)}")

def listar_diretorio() -> None:
    """
    Lista o conteúdo do diretório atual.
    
    Raises:
        OSError: Se ocorrer erro ao acessar o diretório
    """
    try:
        # Usando Path para uma abordagem mais moderna e segura
        diretorio_atual = Path(".")
        
        print("Conteúdo do diretório atual:")
        for item in diretorio_atual.iterdir():
            print(item.name)
    except OSError as e:
        print(f"Erro ao listar diretório: {str(e)}")

def main() -> None:
    """Função principal que processa argumentos da linha de comando."""
    parser = argparse.ArgumentParser(
        description="Programa para gerenciar diretórios",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    
    parser.add_argument(
        "-c", "--criar",
        help="Criar um novo diretório",
        metavar="NOME_DIR"
    )
    parser.add_argument(
        "-l", "--listar",
        action="store_true",
        help="Listar conteúdo do diretório atual"
    )
    
    args = parser.parse_args()
    
    # Processamento dos argumentos
    if args.criar:
        criar_diretorio(args.criar)
    elif args.listar:
        listar_diretorio()
    else:
        parser.print_help()

if __name__ == "__main__":
    main()